packages/compass-data-modeling/src/store/relationships.ts

```diff
@@ -0,0 +1,231 @@
+          return [];
+        });
+      if (!matchingDoc) {
+        return [];
+      }
+      const matchingPropertyPaths = findPropertyPathsMatchingDocs(
+        matchingDoc,
+        schemaPropertyPaths,
+        idsToMatch
+      );
+      return matchingPropertyPaths.map((propPath) => {
+        return [
+          { ns: coll.ns, fields: propPath, cardinality: 1 },
+          { ns: namespace, fields: ['_id'], cardinality: 1 },
+        ] as const;
+      });
+    })
+  );
+  return relationships.flat();
+}
+
+/**
+ * An algorithm that tries to automatically infer possible relationships
+ * starting from a local collection and assuming that any fields that are either
+ * indexed or ObjectIds are pointing to some other foreign collection in the
+ * database
+ */
+export async function inferLocalToForeignRelationshipsForCollection(
+  namespace: string,
+  schema: MongoDBJSONSchema,
+  sampleDocs: Document[],
+  collections: { ns: string; schema: MongoDBJSONSchema }[],
+  dataService: DataService,
+  abortSignal?: AbortSignal
+): Promise<Relationship['relationship'][]> {
+  const indexes = await dataService.indexes(namespace).catch(() => {
+    return null;
+  });
+  const indexesFields = new Set(
+    indexes?.flatMap((index) => {
+      return index.fields.flatMap((field) => {
+        return typeof field.value === 'number' ? field.field : [];
+      });
+    })
+  );
+  const schemaPropertyPaths: string[][] = [];
+  traverseMongoDBJSONSchema(schema, (schema, path) => {
+    if (
+      path[0] !== '_id' &&
+      // Assuming existing index indicate that the field is pointing to an id in
+      // a different namespace.
+      (indexesFields.has(
+        // When getting index information we don't know whether the field has
+        // dots in the name or if it's a nested object / array field. To keep
+        // the check simple we just stringify the detailed schema path for
+        // comparison
+        path.join('.')
+      ) ||
+        // Assuming that ObjectIds might point to some _id in a different
+        // namespace
+        schema.bsonType === 'objectId')
+    ) {
+      schemaPropertyPaths.push(path);
+    }
+  });
+  const maybeRelationships = schemaPropertyPaths.flatMap((path) => {
+    const sampleValues = sampleDocs.flatMap((doc) => {
+      return getValuesFromPath(doc, path);
+    });
+    return collections.map(({ ns }) => {
+      return {
+        foreignNamespace: ns,
+        localPropertyPath: path,
+        sampleValues,
+      };
+    });
+  });
+  const relationships = await Promise.all(
+    maybeRelationships.map(
+      async ({
+        foreignNamespace,
+        localPropertyPath,
+        sampleValues,
+      }): Promise<Relationship['relationship'] | null> => {
+        if (foreignNamespace === namespace) {
+          return null;
+        }
+        const [matchingDoc] = await dataService
+          .find(
+            foreignNamespace,
+            { _id: { $in: sampleValues as any } },
```

Comment:
Should probably take foreign namespace _id type into account, if they are completely mismatched with what we think is the foreign keys we found in local colleciton, we can just skip fetching